Sources/NIOCore/Codec.swift

```diff
@@ -812,3 +857,21 @@ extension MessageToByteHandler {
         }
     }
 }
+
+// MARK: ByteToMessageHandler: RemovableChannelHandler
+extension MessageToByteHandler: RemovableChannelHandler {
+    public func removeHandler(context: ChannelHandlerContext, removalToken: ChannelHandlerContext.RemovalToken) {
+        precondition(!self.isBeingRemoved)
+        self.isBeingRemoved = true
+        context.eventLoop.execute {
+            assert(!self.state.isOperational, "illegal state: \(self.state)")
+            if self.isBeingRemoved {
+                self.isBeingRemoved = false
+            } else {
+                assertionFailure("illegal removal state: \(self.isBeingRemoved)")
+            }
+            // this is necessary as it'll complete the promise.
+            context.leavePipeline(removalToken: removalToken)
+        }
+    }
```

Comment:
I think we can simplify this still further.

`isBeingRemoved` and the `eventLoop.execute` call both exist to try to unwind the call stack and avoid Law of Exclusivity violations. We won't have to do that here: because we don't make it possible to perform outcalls from the MessageToByteEncoder it's not likely that they will happen.

As a result, we can avoid the dance with setting `isBeingRemoved` and then calling `encodeLast` from `write`. Instead, we can simply drop the `data:` argument from `encodeLast` and synchronously call it here, sending any data, before synchronously calling `leavePipeline`.

Does that make sense?

Reply:
Should encodeLast still get passed the buffer and/or be throwing ?
Since decodeLast will be called without data should it's default implementation do nothing ?

Reply:
Yes, it should be passed the buffer and be throwing.

Yes, the default implementation should do nothing.

Reply:
Hi, I'm not sure if you saw my last commit and/or if there anything you want me to change. 
As recommended here and on the forum I refactored my code to use state instead of removing my handlers but if you are still interested into merging this I could do any requested changes.

Reply:
Sorry for the delay, I was on vacation.